src/ert_storage/endpoints/records.py

```diff
@@ -555,70 +655,76 @@ def get_ensemble_responses(
     }
 
 
-def _get_dataframe(
-    content: Any, record: ds.Record, realization_index: Optional[int]
+def _get_record_dataframe(
+    record: ds.Record,
+    realization_index: Optional[int],
+    label: Optional[str],
 ) -> pd.DataFrame:
-    data = pd.DataFrame(content)
+    type_ = record.record_info.record_type
+    if type_ != ds.RecordType.f64_matrix:
+        raise exc.ExpectationError("Non matrix record not supported")
+
     labels = record.f64_matrix.labels
-    if labels is not None and realization_index is None:
-        data.columns = labels[0]
-        data.index = labels[1]
-    elif labels is not None and realization_index is not None:
-        # The output is such that rows are realizations. Because
-        # `content` is a 1d list in this case, it treats each element as
-        # its own row. We transpose the data so that all of the data
-        # falls on the same row.
-        data = data.T
-        data.columns = labels[0]
-        data.index = [realization_index]
+    if labels is not None and label is not None and label not in labels[0]:
+        raise exc.UnprocessableError(f"Record label '{label}' not found!")
+
+    if realization_index is None or record.realization_index is not None:
+        matrix_content = record.f64_matrix.content
+    elif record.realization_index is None:
+        matrix_content = record.f64_matrix.content[realization_index]
+    if not isinstance(matrix_content[0], List):
+        matrix_content = [matrix_content]
+
+    if labels is not None and label is not None:
+        lbl_idx = labels[0].index(label)
+        data = pd.DataFrame([[c[lbl_idx]] for c in matrix_content])
+    else:
+        data = pd.DataFrame(matrix_content)
+
+    if labels is not None:
+        data.columns = labels[0] if label is None else [label]
+        if realization_index is None and record.realization_index is not None:
+            data.index = [record.realization_index]
+        else:
+            data.index = labels[1] if realization_index is None else [realization_index]
```

Comment:
The logic in this block is rather confusing. Is it possible to simplify or add a comment about whats going on?

Reply:
I will try to simplify and also add a comment the behaviour here is quite complex